internal/app/db.go

```diff
@@ -0,0 +1,331 @@
+package app
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/schollz/progressbar/v3"
+	"github.com/tablelandnetwork/basin-cli/pkg/pgrepl"
+	"golang.org/x/exp/slog"
+)
+
+// Column represents a column in a table being replicated.
+type Column struct {
+	Name, Typ         string
+	IsNull, IsPrimary bool
+}
+
+// DBManager manages a duckdb database.
+type DBManager struct {
+	db        *sql.DB
+	dbDir     string
+	dbFname   string
+	table     string
+	createdAT time.Time
+	cols      []Column
+	winSize   time.Duration
+	uploader  *BasinUploader
+}
+
+// NewDBManager creates a new DBManager.
+func NewDBManager(dbDir, table string, cols []Column, winSize time.Duration, uploader *BasinUploader) *DBManager {
+	return &DBManager{
+		dbDir:    dbDir,
+		table:    table,
+		cols:     cols,
+		winSize:  winSize,
+		uploader: uploader,
+	}
+}
+
+// Close closes the current db.
+func (dbm *DBManager) Close() error {
+	return dbm.db.Close()
+}
+
+// queryFromWAL creates a query for a WAL TX records.
+// (todo): error handling and remapping of types.
+func (dbm *DBManager) queryFromWAL(tx *pgrepl.Tx) string {
+	var queries []string
+	for _, r := range tx.Records {
+		columnNames := []string{}
+		vals := []string{}
+		for _, c := range r.Columns {
+			columnNames = append(columnNames, c.Name)
+			vals = append(vals, strings.ReplaceAll(string(c.Value), "\"", ""))
+		}
+
+		cols := strings.Join(columnNames, ", ")
+		valsStr := strings.Join(vals, ", ")
+		query := fmt.Sprintf(
+			"insert into %s (%s) values (%s) \n",
+			r.Table, cols, valsStr,
+		)
+		queries = append(queries, query)
+	}
+
+	// batch all insert queries into one
+	return strings.Join(queries, "\n")
+}
+
+func (dbm *DBManager) replace(ctx context.Context) error {
+	// Export current db to a parquet file at a given path
+	exportAt := path.Join(dbm.dbDir, dbm.dbFname) + ".parquet"
+	if err := dbm.Export(ctx, exportAt); err != nil {
+		return err
+	}
+
+	oldDBFname := dbm.dbFname
+
+	// Close current db
+	slog.Info("closing current db")
+	if err := dbm.Close(); err != nil {
+		return err
+	}
+
+	// Upload the exported parquet file
+	if err := dbm.UploadAt(ctx, exportAt); err != nil {
+		fmt.Println("upload error, skipping", "err", err)
+	}
+
+	// Cleanup the previous db and wal files
+	oldDBPath := path.Join(dbm.dbDir, oldDBFname)
+	if err := dbm.cleanup(oldDBPath); err != nil {
+		return fmt.Errorf("cleanup: %s", err)
+	}
+
+	// Create a new db
+	db, err := dbm.NewDB()
```

Comment:
this pattern is weird

```
db, err := dbm.NewDB()
...
dbm.db = db
```

you're returning a `db` and setting into in the `dbm` later. Why not simply do the `dbm.db = db` inside `NewDB`?

Reply:
also, if every time you call `NewDB` you have to call `Setup`, you can make `Setup` private and be called inside `NewDb`.

Reply:
makes sense 👍